pkg/scalers/elasticsearch_scaler.go

```diff
@@ -0,0 +1,276 @@
+	if err != nil {
+		elasticsearchLog.Error(err, fmt.Sprintf("Found error when creating client: %s", err))
+		return nil, err
+	}
+
+	_, err = esClient.Info()
+	if err != nil {
+		elasticsearchLog.Error(err, fmt.Sprintf("Found error when pinging search engine: %s", err))
+		return nil, err
+	}
+	return esClient, nil
+}
+
+func (s *elasticsearchScaler) Close(ctx context.Context) error {
+	return nil
+}
+
+// IsActive returns true if there are pending messages to be processed
+func (s *elasticsearchScaler) IsActive(ctx context.Context) (bool, error) {
+	messages, err := s.getQueryResult(ctx)
+	if err != nil {
+		elasticsearchLog.Error(err, fmt.Sprintf("Error inspecting elasticsearch: %s", err))
+		return false, err
+	}
+	return messages > 0, nil
+}
+
+// getQueryResult returns result of the scaler query
+func (s *elasticsearchScaler) getQueryResult(ctx context.Context) (int, error) {
+	// Build the request body.
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(buildQuery(s.metadata)); err != nil {
+		elasticsearchLog.Error(err, "Error encoding query: %s", err)
+	}
+
+	// Run the templated search
+	res, err := s.esClient.SearchTemplate(
+		&body,
+		s.esClient.SearchTemplate.WithIndex(s.metadata.indexes...),
+		s.esClient.SearchTemplate.WithContext(ctx),
+	)
+	if err != nil {
+		elasticsearchLog.Error(err, fmt.Sprintf("Could not query elasticsearch: %s", err))
+		return 0, err
+	}
+
+	defer res.Body.Close()
+	b, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return 0, err
+	}
+	v, err := getValueFromSearch(b, s.metadata.valueLocation)
+	if err != nil {
+		return 0, err
+	}
+	return v, nil
+}
+
+func buildQuery(metadata *elasticsearchMetadata) map[string]interface{} {
+	parameters := map[string]interface{}{}
+	for _, p := range metadata.parameters {
+		if p != "" {
+			kv := splitAndTrimBySep(p, ":")
+			parameters[kv[0]] = kv[1]
+		}
+	}
+	query := map[string]interface{}{
+		"id": metadata.searchTemplateName,
+	}
+	if len(parameters) > 0 {
+		query["params"] = parameters
+	}
+	return query
+}
+
+func getValueFromSearch(body []byte, valueLocation string) (int, error) {
+	r := gjson.GetBytes(body, valueLocation)
+	errorMsg := "valueLocation must point to value of type number but got: '%s'"
+	if r.Type == gjson.String {
+		q, err := strconv.Atoi(r.String())
+		if err != nil {
+			return 0, fmt.Errorf(errorMsg, r.String())
+		}
+		return q, nil
+	}
+	if r.Type != gjson.Number {
+		return 0, fmt.Errorf(errorMsg, r.Type.String())
+	}
+	return int(r.Num), nil
+}
+
+// GetMetricSpecForScaling returns the MetricSpec for the Horizontal Pod Autoscaler
+func (s *elasticsearchScaler) GetMetricSpecForScaling(context.Context) []v2beta2.MetricSpec {
+	targetValue := resource.NewQuantity(int64(s.metadata.targetValue), resource.DecimalSI)
+
+	externalMetric := &v2beta2.ExternalMetricSource{
+		Metric: v2beta2.MetricIdentifier{
+			Name: GenerateMetricNameWithIndex(s.metadata.targetValue, s.metadata.metricName),
```

Comment:
You should call only once to `GenerateMetricNameWithIndex`. This index is not related with any inside the scaler, it's related with the trigger index inside the ScaledObject to avoid conflicts with duplicated names in the same ScaledObject.
You could call to the function here or in the constructor like you are doing but don't do it both, because the final metric name with current code will be like `s{s.metadata.targetValue}-s{config.ScalerIndex}-elasticsearch-%s`. It should be like  `s{config.ScalerIndex}-elasticsearch-%s`

Reply:
yes, My bad ; it's fixed and I've just add the proper unit test